packages/lexicon/src/validators/primitives.ts

```diff
@@ -197,9 +197,14 @@ export function string(
     }
   }
 
+  // Lazily calculated and reused between checks.
+  let cachedUtf8Len: number | undefined
```

Comment:
The "caching" logic could maybe be more readable like:

```typescript
if (typeof def.maxLength ==='number' || typeof def.minLength ==='number') {
    const len = utf8Len(value)
    if (typeof def.maxLength ==='number') {
        if (len > def.maxLength) {
            ....
        }
    }
    if (typeof def.minLength ==='number') {
        ...
    }
}
```
and similar for the grapheme value.

but not sure what idiomatic typescript style would be here. not a big deal either way.

Reply:
ah yea that's much nicer